src/base/printer.hpp

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * printer.hpp - printing classes
+ */
+
+#ifndef _PRINTER_HPP_
+#define _PRINTER_HPP_
+
+#include "string_buffer.hpp"
+#include <string>
+
+/*
+ * print_output : interface class to print contents
+ * operator () : template variadic operator to allows printf-like syntax
+ *               it automaticaly calls virtual flush () method which needs to be implemented in specialized class
+ */
+class print_output
+{
+  private:
+    std::string context_name;
```

Comment:
Maybe context_name/exec_name belongs to extract_context rather than print_output?